File: python/mirage/commands/builtin/s3/ls/ls.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.s3 import S3Accessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.s3._provision import metadata_provision
from mirage.commands.builtin.utils.formatting import _human_size
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.s3.glob import resolve_glob
from mirage.core.s3.readdir import readdir
from mirage.core.s3.stat import stat
from mirage.io.types import ByteSource, IOResult
from mirage.types import FileType, PathSpec


def _get_extension(path: str) -> str | None:
    dot = path.rfind(".")
    if dot == -1 or "/" in path[dot:]:
        return None
    return path[dot:]


async def _ls_entries(
    accessor: S3Accessor,
    path: PathSpec,
    all_files: bool,
    sort_by: str,
    reverse: bool,
    recursive: bool,
    list_dir: bool,
    warnings: list[str],
    index: IndexCacheStore = None,
):
    if list_dir:
        s = await stat(accessor, path, index)
        return [s]

    try:
        entries = await readdir(accessor, path, index)
    except (FileNotFoundError, ValueError) as exc:
        warnings.append(f"ls: cannot access '{path.original}': {exc}")
        return []

    stats = []
    for entry in entries:
        try:
            e_spec = PathSpec(original=entry,
                              directory=entry,
                              resolved=False,
                              prefix=path.prefix)
            s = await stat(accessor, e_spec, index)
        except (FileNotFoundError, ValueError) as exc:
            warnings.append(f"ls: cannot access '{entry}': {exc}")
            continue
        if not all_files and s.name.startswith("."):
            continue
        stats.append(s)

    if sort_by == "time":
        stats.sort(key=lambda s: s.modified or "", reverse=not reverse)
    elif sort_by == "size":
        stats.sort(key=lambda s: s.size or 0, reverse=not reverse)
    else:
        stats.sort(key=lambda s: s.name, reverse=reverse)

    if recursive:
        sub_entries = []
        for s in stats:
            sub_entries.append(s)
            if s.type == FileType.DIRECTORY:
                entry_path = path.child(s.name)
                entry_spec = PathSpec(original=entry_path,
                                      directory=entry_path,
                                      resolved=False,
                                      prefix=path.prefix)
                sub = await _ls_entries(accessor, entry_spec, all_files,
                                        sort_by, reverse, recursive, False,
                                        warnings, index)
                sub_entries.extend(sub)
        return sub_entries

    return stats


@command("ls", resource="s3", spec=SPECS["ls"], provision=metadata_provision)
async def ls(
    accessor: S3Accessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: bytes | None = None,
    args_l: bool = False,
    args_1: bool = False,
    a: bool = False,
    A: bool = False,
    h: bool = False,
    t: bool = False,
    S: bool = False,
    r: bool = False,
    R: bool = False,
    d: bool = False,
    F: bool = False,
    filetype_fns: dict | None = None,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    paths = await resolve_glob(accessor, paths, index)
    targets = [p for p in paths]
    all_files = a or A
    sort_by = "name"
    if t:
        sort_by = "time"
    elif S:
        sort_by = "size"
    warnings: list[str] = []
    results: list[str] = []
    for p in targets:
        p.prefix if isinstance(p, PathSpec) else ""
        try:
            entries = await _ls_entries(
                accessor,
                p,
                all_files,
                sort_by,
                r,
                R,
                d,
                warnings,
                index,
            )
        except (FileNotFoundError, ValueError) as exc:
            warnings.append(f"ls: cannot access '{p.original}': {exc}")
            continue
        if args_l and not args_1:
            for e in entries:
                ext = _get_extension(e.name)
                if filetype_fns and ext in filetype_fns:
                    try:
                        fn = filetype_fns[ext]
                        path_for_entry = (p.original if isinstance(
                            p, PathSpec) else p).rstrip("/") + "/" + e.name
                        stdout, _io = await fn(
                            accessor,
                            [path_for_entry],
                            args_l=True,
                        )
                        if stdout:
                            if isinstance(stdout, bytes):
                                results.append(stdout.decode(errors="replace"))
                            else:
                                chunks = [chunk async for chunk in stdout]
                                results.append(
                                    b"".join(chunks).decode(errors="replace"))
                            continue
                    except Exception:
                        pass
                size_str = _human_size(e.size or 0) if h else str(e.size or 0)
                line = (f"{e.type or '-'}\t{size_str}"
                        f"\t{e.modified or ''}\t{e.name}")
                results.append(line)
        else:
            for e in entries:
                is_dir = F and e.type == FileType.DIRECTORY
                name = e.name + "/" if is_dir else e.name
                results.append(name)
    stderr = "\n".join(warnings).encode() if warnings else None
    exit_code = 1 if warnings and not results else 0
    output = ("\n".join(results) + "\n").encode() if results else b""
    return output, IOResult(stderr=stderr, exit_code=exit_code)

File: typescript/packages/core/src/commands/builtin/gdrive/ls.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GDriveAccessor } from '../../../accessor/gdrive.ts'
import { resolveGlob } from '../../../core/gdrive/glob.ts'
import { readdir as gdriveReaddir } from '../../../core/gdrive/readdir.ts'
import { stat as gdriveStat } from '../../../core/gdrive/stat.ts'
import { IOResult, type ByteSource } from '../../../io/types.ts'
import type { FileStat } from '../../../types.ts'
import { FileType, PathSpec, ResourceName } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { humanSize } from '../utils/formatting.ts'
import { metadataProvision } from './provision.ts'

const ENC = new TextEncoder()

async function lsEntries(
  accessor: GDriveAccessor,
  path: PathSpec,
  allFiles: boolean,
  sortBy: 'name' | 'size',
  reverse: boolean,
  recursive: boolean,
  listDir: boolean,
  warnings: string[],
  indexCache: CommandOpts['index'],
): Promise<FileStat[]> {
  if (listDir) {
    const s = await gdriveStat(accessor, path, indexCache ?? undefined)
    return [s]
  }
  let entries: string[]
  try {
    entries = await gdriveReaddir(accessor, path, indexCache ?? undefined)
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err)
    warnings.push(`ls: cannot access '${path.original}': ${msg}`)
    return []
  }
  const stats: FileStat[] = []
  for (const entry of entries) {
    try {
      const eSpec = new PathSpec({
        original: entry,
        directory: entry,
        resolved: false,
        prefix: path.prefix,
      })
      const s = await gdriveStat(accessor, eSpec, indexCache ?? undefined)
      if (!allFiles && s.name.startsWith('.')) continue
      stats.push(s)
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err)
      warnings.push(`ls: cannot access '${entry}': ${msg}`)
    }
  }
  if (sortBy === 'size') {
    stats.sort((a, b) => (a.size ?? 0) - (b.size ?? 0))
    if (!reverse) stats.reverse()
  } else {
    stats.sort((a, b) => a.name.localeCompare(b.name))
    if (reverse) stats.reverse()
  }
  if (recursive) {
    const subEntries: FileStat[] = []
    for (const s of stats) {
      subEntries.push(s)
      if (s.type === FileType.DIRECTORY) {
        const childPath = path.child(s.name)
        const childSpec = new PathSpec({
          original: childPath,
          directory: childPath,
          resolved: false,
          prefix: path.prefix,
        })
        const sub = await lsEntries(
          accessor,
          childSpec,
          allFiles,
          sortBy,
          reverse,
          recursive,
          false,
          warnings,
          indexCache,
        )
        subEntries.push(...sub)
      }
    }
    return subEntries
  }
  return stats
}

async function lsCommand(
  accessor: GDriveAccessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const resolved = await resolveGlob(accessor, paths, opts.index ?? undefined)
  const targets: PathSpec[] =
    resolved.length > 0
      ? resolved
      : [
          new PathSpec({
            original: opts.cwd,
            directory: opts.cwd,
            resolved: false,
            prefix: opts.mountPrefix ?? '',
          }),
        ]
  const long = opts.flags.args_l === true && opts.flags.args_1 !== true
  const allFiles = opts.flags.a === true || opts.flags.A === true
  const human = opts.flags.h === true
  const reverse = opts.flags.r === true
  const recursive = opts.flags.R === true
  const listDir = opts.flags.d === true
  const classify = opts.flags.F === true
  const sortBy: 'name' | 'size' = opts.flags.S === true ? 'size' : 'name'
  const warnings: string[] = []
  const results: string[] = []
  for (const p of targets) {
    let entries: FileStat[]
    try {
      entries = await lsEntries(
        accessor,
        p,
        allFiles,
        sortBy,
        reverse,
        recursive,
        listDir,
        warnings,
        opts.index,
      )
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err)
      warnings.push(`ls: cannot access '${p.original}': ${msg}`)
      continue
    }
    if (long) {
      for (const e of entries) {
        const sizeStr = human ? humanSize(e.size ?? 0) : String(e.size ?? 0)
        results.push(`${e.type ?? '-'}\t${sizeStr}\t${e.modified ?? ''}\t${e.name}`)
      }
    } else {
      for (const e of entries) {
        const isDir = classify && e.type === FileType.DIRECTORY
        const name = isDir ? e.name + '/' : e.name
        results.push(name)
      }
    }
  }
  const stderr = warnings.length > 0 ? ENC.encode(warnings.join('\n')) : null
  const exitCode = warnings.length > 0 && results.length === 0 ? 1 : 0
  const out: ByteSource = ENC.encode(results.join('\n'))
  return [out, new IOResult({ stderr, exitCode })]
}

export const GDRIVE_LS = command({
  name: 'ls',
  resource: ResourceName.GDRIVE,
  spec: specOf('ls'),
  fn: lsCommand,
  provision: metadataProvision,
})
